kubeadm: Implement support for symbolic labels in --use-kubernetes-version

kad opened the pull request:
**What this PR does / why we need it**: This patch adds "stable", "latest" and other labels as valid versions in "kubeadm init --use-kubernetes-version" flag.

Now, defaults can be pointing to "stable" and users will always get
latest available stable build of Kubernetes via kubeadm.
There is no need anymore to hardcode version string inside kubeadm
binary.

 It is also possible to use labels like "latest" or point to exact
 branch: "stable-1.4"

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

**Release note**:
```release-note
- kubeadm init --use-kubernetes-version now understands "stable","latest", "stable-1.4" and other labels as valid version. It will fetch actual build number from release servers. 
```

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/37222)
<!-- Reviewable:end -->

kad committed:
Implement support for symbolic labels in --use-kubernetes-version

Now, defaults can be pointing to "stable" and users will always get
latest available stable build of Kubernetes via kubeadm.
There is no need anymore to hardcode version string inside kubeadm
binary.

It is also possible to use labels like "latest" or point to exact
branch: "stable-1.4"

kad commented:
@kubernetes/sig-cluster-lifecycle

dims commented:
@k8s-bot ok to test

luxas commented on cmd/kubeadm/app/util/version.go:
```diff
@@ -0,0 +1,89 @@
+	} else if kubeReleaseLabelRegex.MatchString(version) {
+		url = fmt.Sprintf("%s/%s.txt", kubeReleaseBucketURL, version)
+	} else if kubeCiReleaseLabelRegex.MatchString(version) {
+		url = fmt.Sprintf("%s/%s.txt", kubeDevReleaseBucketURL, version)
```
Example of https://storage.googleapis.com/kubernetes-release-dev/ci/latest.txt is `v1.6.0-alpha.0.760+4dd0320db7cd3d`, but no CI images are actually pushed for the server components https://gcr.io/v2/google-containers/kube-apiserver-amd64/tags/list

So at least we can't use the `ci/*` versions right now because there aren't images for them (unless we build them ourselves, but that's out of scope for now at least)

kad replied:
Hm. didn't notice that. I had impression that images are also published in similar way how actual builds are published to gs://kubernetes-release-dev/ci/$build/ Ok, will remove ci/* for now.

luxas commented on cmd/kubeadm/app/util/version.go:
```diff
@@ -0,0 +1,89 @@
+
+var (
+	kubeReleaseRegex        = regexp.MustCompile(`^v(0|[1-9][0-9]*)\.(0|[1-9][0-9]*)\.(0|[1-9][0-9]*)([-0-9a-zA-Z_\.+]*)?$`)
+	kubeReleaseLabelRegex   = regexp.MustCompile(`^release/[-\w_\.]+`)
```
Could we just allow `release/(stable|latest)-[\.\d]+` or something here?
Right now I feel that for example `release/foobar-1.7` could be passed to it just fine and the validation will pass

kad replied:
validation will pass, but then it will fail to fetch corresponding txt file.
I'd rather not hardcode labels here, so build team can introduce labels more easily, without need to modify kubeadm code.

luxas commented on cmd/kubeadm/app/util/version.go:
```diff
@@ -0,0 +1,89 @@
+	} else if kubeCiReleaseLabelRegex.MatchString(version) {
+		url = fmt.Sprintf("%s/%s.txt", kubeDevReleaseBucketURL, version)
+	} else {
+		return "", fmt.Errorf("Error in version string %q", version)
```
`The version string didn't match the allowed patterns: [pattern1] [pattern2]`

luxas commented on cmd/kubeadm/app/util/version.go:
```diff
@@ -0,0 +1,89 @@
+func fetchVersion(url string) (string, error) {
+	resp, err := http.Get(url)
+	if err != nil {
+		return "", err
```
Wrap the error here with something more useful, e.g. `The placeholder version string that was specified %s wasn't resolvable to an actual release. Please try again with internet connection or with another placeholder. Error: %v`

luxas commented on cmd/kubeadm/app/util/version.go:
```diff
@@ -0,0 +1,89 @@
+	defer resp.Body.Close()
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return "", err
```
Wrap the error here with something more useful, e.g. `The placeholder version string that was specified %s wasn't resolvable to an actual release. Please try again with internet connection or with another placeholder. Error: %v`

luxas commented on cmd/kubeadm/app/cmd/init.go:
```diff
@@ -193,6 +193,16 @@ func NewInit(cfgPath string, cfg *kubeadmapi.MasterConfiguration, skipPreFlight
+	if err != nil {
+		return nil, err
+	}
+	if ver != cfg.KubernetesVersion {
```
No need to do a separate check here, just do
```
cfg.KubernetesVersion = ver
fmt.Printf("Using Kubernetes version: %d", cfg.KubernetesVersion)
```

That's also more readable

luxas reviewed (commented):
LGTM overall, but please omit the `ci/*` labels and fix the nits

kad commented:
Uploaded new version with addressed feedback and removed completely support of ci/* labels, as they are not really usable.

kad commented:
Hm. this test failure seems to be caused by slowness in test cluster, not by PR itself.

luxas commented:
@k8s-bot cvm gce e2e test this

luxas commented:
> Hm. this test failure seems to be caused by slowness in test cluster, not by PR itself.

@kad That's correct. We call it a flake.

LGTM

kad commented:
@luxas @saad-ali would you consider this to be part of 1.5 ? I think it's simple but very useful functionality.

luxas commented:
No, kubeadm won't release a new version before v1.5, and it will be splitted out just after the master branch opens for v1.6. It just don't follow the main releasing scheme yet.

See the release notes doc here http://kubernetes.io/docs/admin/kubeadm/ and the issue here: https://github.com/kubernetes/kubernetes/issues/34884

We will, however, release a stable version (third one) from the master branch just after master has opened next week, where I will include this PR in that release.

bgrant0607 commented:
@luxas Please do not add/change priority labels on PRs.

cc @eparis @saad-ali @kubernetes/sig-cluster-lifecycle

luxas commented:
@bgrant0607 Please see: https://github.com/kubernetes/contrib/pull/2084, and we'll continue the discussion there or on Slack.
I'm happy to change method, but I didn't get a no when discussing it on Slack earlier today.
At least we have to find some method that works for getting the v1.5 and kubeadm PRs in as quickly as possible.

mikedanese commented on cmd/kubeadm/app/util/version.go:
```diff
@@ -0,0 +1,74 @@
+//  latest      (latest release, including alpha/beta)
+//  latest-1    (latest release in 1.x, including alpha/beta)
+//  latest-1.0  (and similarly 1.1, 1.2, 1.3, ...)
+func KubernetesReleaseVersion(version string) (string, error) {
```
Can we get a test for this?

luxas replied:
+1

marun replied:
I've taken a look at existing testing around ``http.Get`` and it often seems to involve spinning up local http servers (even for unit tests).  Another approach would be refactoring the function so that http.Get could be replaced in testing by fake that could return whatever was necessary to exercise the code paths of the function.  Would one of those approaches be appropriate here?

kad replied:
@marun I think I found couple of ways to do mocking of those http calls. One option would be to make URL as var instead of constant and update it, second way to replace default transport during test execution. Probably go with second way for now.

luxas commented on cmd/kubeadm/app/util/version.go:
```diff
@@ -0,0 +1,74 @@
+		// Re-validate received version and return.
+		return KubernetesReleaseVersion(strings.Trim(string(body), " \t\n"))
+	}
+	return "", fmt.Errorf("Error: version %q doesn't match patterns for neither semantic version nor labels (stable, latest, ...)", version)
```
We should fallback to a well-known version if internet isn't available, so if the user didn't set the flag, it should default to `v1.4.4` or something.

If the user had the flag set, it should fail if it was invalid indeed.

Sorry, I thought I had commented that already, but somehow it seems I hadn't

luxas reviewed (changes requested):
One thing that has to be fixed yet...

kad committed:
Added test case for KubernetesReleaseVersion

kad committed:
Fallback to known good stable version in case of network errors.

Hardcoded known stable version will be returned if user
didn't request specific version and kubeadm for some reason
not able to fetch latest stable information from release servers.

For now, fallback version is v1.4.6

kad commented:
PR updated with test suite and requested additional functionality.

kad commented:
Looks like flake :( can someone restart ?

zmerlynn commented:
@kad: You can, just parrot the command back as it requested. However, it looks like we dropped one message here.

@k8s-bot gci gke e2e test this
@k8s-bot cvm gce e2e test this

kad commented:
@zmerlynn I had impression that those commands are ok only for org members

zmerlynn commented:
@k8s-bot verify test this
@kad If I understand correctly, you can still issue test commands (per https://github.com/kubernetes/kubernetes/pull/37222#issuecomment-261961314), but I'm not entirely sure. This isn't well documented, I'm sorry.

luxas commented:
Thanks, LGTM
